src/wrapper/java_vm/vm.rs

```diff
@@ -132,9 +190,128 @@ impl<'a> Deref for AttachGuard<'a> {
 }
 
 impl<'a> Drop for AttachGuard<'a> {
+    fn drop(&mut self) {
+        if self.should_detach {
+            InternalAttachGuard::clear_tls();
+        }
+    }
+}
+
+#[derive(PartialEq)]
+enum ThreadType {
+    Normal,
+    Daemon,
+}
+
+#[derive(Debug)]
+struct InternalAttachGuard {
+    java_vm: *mut sys::JavaVM,
+}
+
+impl InternalAttachGuard {
+    fn create_and_attach(
+        java_vm: *mut sys::JavaVM,
+        ty: ThreadType,
+    ) -> Result<*mut sys::JNIEnv> {
+        let guard = InternalAttachGuard::new(java_vm);
+        let env_ptr = unsafe {
+            if ty == ThreadType::Daemon {
+                guard.attach_current_thread_as_daemon()?
+            } else {
+                guard.attach_current_thread()?
+            }
+        };
+
+        Self::fill_tls(guard);
+
+        Ok(env_ptr)
+    }
+
+    fn new(java_vm: *mut sys::JavaVM) -> Self {
+        Self {
+            java_vm,
+        }
+    }
+
+    /// Stores guard in thread local storage.
+    fn fill_tls(guard: InternalAttachGuard) {
+        THREAD_ATTACH_GUARD.with(move |f| {
+            *f.borrow_mut() = Some(guard);
+        });
+    }
+
+    /// Clears thread local storage, drops InternalAttachGuard and causes detach.
+    fn clear_tls() {
+        THREAD_ATTACH_GUARD.with(move |f| {
+            *f.borrow_mut() = None;
+        });
+    }
+
+    unsafe fn attach_current_thread(&self) -> Result<*mut sys::JNIEnv> {
+        let mut env_ptr = ptr::null_mut();
+        let res = java_vm_unchecked!(
+            self.java_vm,
+            AttachCurrentThread,
+            &mut env_ptr,
+            ptr::null_mut()
+        );
+        jni_error_code_to_result(res)?;
+
+        ATTACHED_THREADS.fetch_add(1, Ordering::SeqCst);
+
+        debug!("Attached thread {:?}. Name: {:?}. {} threads attached",
```

Comment:
Nit (here and elsewhere): would the following format be more human-readable:

`Attached thread foo (id=1234).`

?

Reply:
> Nit (here and elsewhere): would the following format be more human-readable:
> 
> `Attached thread foo (id=1234).`
> 
> ?

Thread is not required to have a name, that's why `.name()` returns `Option<>`. I guess we could use "Attached thread ["NAME_" or ""] with id: ID".

Reply:
Btw, Debug implementation for ThreadId really sucks - it will be printed as `ThreadId(<number>)`, and it seems like we can't make it nicer

Reply:
It's ok — you don't have to put `id=` :-)